File: protocol/ts/src/binary.ts
import { MAX_UINT64 } from "./constants.js";
import { ProtocolError } from "./errors.js";

const UUID_PATTERN = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/;
const UUID_V4_PATTERN = /^[0-9a-f]{8}-[0-9a-f]{4}-4[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/;
const UINT64_PATTERN = /^(0|[1-9][0-9]{0,19})$/;
const LEAF_ID_PATTERN = /^[0-9a-f]{8}$/;

export function parseUint64(value: string): bigint {
  if (!UINT64_PATTERN.test(value)) {
    throw new ProtocolError("PROTOCOL_VIOLATION", "uint64 must be canonical decimal text");
  }
  const parsed = BigInt(value);
  if (parsed > MAX_UINT64) {
    throw new ProtocolError("PROTOCOL_VIOLATION", "uint64 exceeds its maximum");
  }
  return parsed;
}

export function formatUint64(value: bigint): string {
  if (value < 0n || value > MAX_UINT64) {
    throw new ProtocolError("PROTOCOL_VIOLATION", "uint64 is out of range");
  }
  return value.toString(10);
}

export function assertLeafId(value: string | null): void {
  if (value !== null && !LEAF_ID_PATTERN.test(value)) {
    throw new ProtocolError("PROTOCOL_VIOLATION", "Leaf ID must be null or eight lowercase hex characters");
  }
}

export function uuidToBytes(uuid: string): Uint8Array {
  if (!UUID_PATTERN.test(uuid)) {
    throw new ProtocolError("PROTOCOL_VIOLATION", "UUID must be lowercase canonical text");
  }
  return Uint8Array.from(Buffer.from(uuid.replaceAll("-", ""), "hex"));
}

export function uuidV4ToBytes(uuid: string): Uint8Array {
  if (!UUID_V4_PATTERN.test(uuid)) {
    throw new ProtocolError("PROTOCOL_VIOLATION", "UUID must be a lowercase UUIDv4");
  }
  return uuidToBytes(uuid);
}

export function bytesToUuid(bytes: Uint8Array): string {
  if (bytes.length !== 16) {
    throw new ProtocolError("PROTOCOL_VIOLATION", "UUID prefix must contain exactly 16 bytes");
  }
  const hex = Buffer.from(bytes).toString("hex");
  return `${hex.slice(0, 8)}-${hex.slice(8, 12)}-${hex.slice(12, 16)}-${hex.slice(16, 20)}-${hex.slice(20)}`;
}

export function assertUuidV4(uuid: string): void {
  uuidV4ToBytes(uuid);
}

export function readUint64BigEndian(bytes: Uint8Array, offset = 0): bigint {
  if (offset < 0 || offset + 8 > bytes.length) {
    throw new ProtocolError("PROTOCOL_VIOLATION", "Truncated uint64");
  }
  return new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength).getBigUint64(offset, false);
}

export function writeUint64BigEndian(value: bigint): Uint8Array {
  formatUint64(value);
  const output = new Uint8Array(8);
  new DataView(output.buffer).setBigUint64(0, value, false);
  return output;
}

File: android/core/protocol/src/main/kotlin/io/github/verybigsad/pimobile/protocol/Binary.kt
package io.github.verybigsad.pimobile.protocol

import java.nio.ByteBuffer
import java.nio.ByteOrder

private val uuidPattern = Regex("^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$")
private val uuidV4Pattern = Regex("^[0-9a-f]{8}-[0-9a-f]{4}-4[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$")
private val uint64Pattern = Regex("^(0|[1-9][0-9]{0,19})$")
private val leafIdPattern = Regex("^[0-9a-f]{8}$")

fun parseUint64(value: String): ULong {
    if (!uint64Pattern.matches(value)) protocolViolation("uint64 must be canonical decimal text")
    return value.toULongOrNull() ?: protocolViolation("uint64 exceeds its maximum")
}

fun formatUint64(value: ULong): String = value.toString()

fun assertLeafId(value: String?) {
    if (value != null && !leafIdPattern.matches(value)) protocolViolation("Leaf ID must be null or eight lowercase hex characters")
}

fun uuidToBytes(uuid: String): ByteArray {
    if (!uuidPattern.matches(uuid)) protocolViolation("UUID must be lowercase canonical text")
    return uuid.replace("-", "").chunked(2).map { it.toInt(16).toByte() }.toByteArray()
}

fun uuidV4ToBytes(uuid: String): ByteArray {
    if (!uuidV4Pattern.matches(uuid)) protocolViolation("UUID must be a lowercase UUIDv4")
    return uuidToBytes(uuid)
}

fun bytesToUuid(bytes: ByteArray): String {
    if (bytes.size != 16) protocolViolation("UUID prefix must contain exactly 16 bytes")
    val hex = bytes.joinToString("") { "%02x".format(it.toInt() and 0xff) }
    return "${hex.substring(0, 8)}-${hex.substring(8, 12)}-${hex.substring(12, 16)}-${hex.substring(16, 20)}-${hex.substring(20)}"
}

fun readUint64BigEndian(bytes: ByteArray, offset: Int = 0): ULong {
    if (offset < 0 || offset + 8 > bytes.size) protocolViolation("Truncated uint64")
    return ByteBuffer.wrap(bytes, offset, 8).order(ByteOrder.BIG_ENDIAN).long.toULong()
}

fun writeUint64BigEndian(value: ULong): ByteArray = ByteBuffer.allocate(8)
    .order(ByteOrder.BIG_ENDIAN)
    .putLong(value.toLong())
    .array()
